ntirpc/rpc/gss_internal.h

```diff
@@ -113,30 +112,17 @@ svc_rpc_gss_data *alloc_svc_rpc_gss_data(void)
 }
 
 static inline void
-unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd,
-		       uint32_t flags)
+unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd)
 {
-	u_int refcnt;
-	bool gd_locked = flags & SVC_RPC_GSS_FLAG_LOCKED;
-
-	refcnt = atomic_dec_uint32_t(&gd->refcnt);
+	mutex_lock(&gd->lock);
```

Comment:
A thread should access gd fields with its own refcnt. The refcnt can't go to zero while it is accessing. If you think a thread could go on manipulating gd fields while another thread tries to destroy, with lock or without lock, we will have issues.

If this object is heavily shared and unref'ed many times, we acquire the lock every time we unref bringing down performance.